_posts/2019-01-26-generic-way-of-handling-loading-saving-and-validationerrors-in-angular.md

```diff
@@ -0,0 +1,410 @@
+): Observable<HttpEvent<any>> {
+  ..
+  return next.handle(req.clone()).pipe(
+    // catch the error
+    catchError(e => {
+      // if bad request > validation erors
+      if (e.status === 400) { 
+        // use the validationErrors setter to update
+        this.httpStatusService.validationErrors = 
+          e.error.validationErrors;
+        // make sure that this result never 
+        // reaches the component
+        return NEVER;
+      }
+      // throw the error back
+      // or put that in the `HttpStatusService` as well ;-)      
+      return throwError(e); 
+    }),
+    finalize(...)
+  );
+}
+```
+
+The entire interceptor class looks like this:
+
+```typescript
+@Injectable({
+    providedIn: 'root'
+})
+export class HttpStatusInterceptor implements HttpInterceptor {
+  private loadingCalls = 0; 
+  private actingCalls = 0; 
+
+  constructor(
+    private httpStatusService: HttpStatusService
+  ) {}
+
+  private changeStatus(v: boolean, method: string): void {
+    if (['POST', 'PUT', 'DELETE', 'PATCH']
+      .indexOf(method) > -1) {
+      v ? this.actingCalls++ : this.actingCalls--;
+      this.httpStatusService.acting = this.actingCalls > 0;
+    } else if (method === 'GET') {
+      v ? this.loadingCalls++ : this.loadingCalls--;
+      this.httpStatusService.loading = this.loadingCalls > 0;
+    }
+  }
+
+  intercept(
+    req: HttpRequest<any>,
+    next: HttpHandler
+  ): Observable<HttpEvent<any>> {
+    return next.handle(req.clone()).pipe(
+      catchError(e => {
+        if (e.status === 400) { 
+          this.httpStatusService.validationErrors = 
+            e.error.validationErrors;
+          return NEVER;
+        }
+        return throwError(e); 
+      }),
+      finalize(() => {
+        this.changeStatus(false, req.method);
+      })
+    );
+  }
+}
+```
+
+When registering the interceptor to the root module, the observables in the `HttpStatusService` will be updated automatically.
+
+So in the rootModule we have to add the following code to the `providers` property of the `@NgModule` decorator:
+
+```typescript
+providers: [
+  ...
+  {
+    provide: HTTP_INTERCEPTORS,
+    multi: true,
+    deps: [HttpStatusService],
+    useClass: HttpStatusInterceptor
+  }
+]
+```
+
+We can now use the `HttpStatusService` in our code as easy as this:
+
+```typescript
+export class UserComponent {
+  loading$ = this.httpStatus.loading;
+  validationErrors$ = this.httpStatus.validationErrors;
+  acting$ = this.httpStatus.loading;
+
+  constructor(
+    private httpStatusService: HttpStatusService) {
+  }
+}
+```
+
+We now have 3 observables that can easily be consumed in the template of the component with the use of the [async](https://angular.io/api/common/AsyncPipe) pipe. Here is an example:
+
+```html
+<my-spinner *ngIf="loading$|async"></my-spinner>
+<my-user-form 
+    [validationErrors]="validationErrors$|async"
+    [disabled]="acting$|async"></my-user-form>
+
+```
+
+## Optimizing with decorators
```

Comment:
Cool, but nice to mention trade-offs off this solution

Reply:
What are the tradeoffs in your opinion?